### PR TITLE
ENH: expose datetime64 conversion functions

### PR DESCRIPTION
**Pull request by jbrockmendel**
These would allow pandas to remove our copy/pasted implementations.

cc @WillAyd @mroeschke anything else we might need?

### CODE REVIEW

File: numpy/core/src/multiarray/_datetime.h [hunk 1]
```diff
@@ -287,8 +286,7 @@ convert_timedelta_to_pyobject(npy_timedelta td, PyArray_DatetimeMetaData *meta);
 /*
  * Converts a datetime based on the given metadata into a datetimestruct
  */
-NPY_NO_EXPORT int
-convert_datetime_to_datetimestruct(PyArray_DatetimeMetaData *meta,
+int convert_datetime_to_datetimestruct(PyArray_DatetimeMetaData *meta,
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/_datetime.h:289]: I don't think this is sufficient, extension modules aren't allowed to link against each other. This needs to go through the function table like the other `NUMPY_API` functions.

Comment (jbrockmendel) [reply numpy/core/src/multiarray/_datetime.h:289]: Thanks, ill see if i can track that down.  Before I do, is this likely to be accepted once working?

Comment (eric-wieser) [reply numpy/core/src/multiarray/_datetime.h:289]: I think we'd need to think about the function names, but in principle it seems sensible to me to expose the `npy_datetimestruct` type and some APIs that consume it.

**Comment by WillAyd** [comment]
For completeness we might want to upstream the timedelta changes as well:

https://github.com/pandas-dev/pandas/blob/707640cc89dd16bff00b62950b3fa27f3b642d5f/pandas/_libs/tslibs/src/datetime/np_datetime_strings.h#L90

Probably in a separate PR, maybe a pre-cursor. The above is used in JSON so I think we wouldn't be able to completely simplify our dependency structure without tacking that as well

**Comment by bashtage** [comment]
@eric-wieser Not sure why this is tagged random.

**Comment by jbrockmendel** [comment]
> Not sure why this is tagged random

@bashtage best guess is that on a previous PR implementing test_cython we adapted the setup from random.tests.test_extending into a fixture and there was an offhand mention of using that pattern in the random tests.

**Comment by jbrockmendel** [comment]
Closing in favor of #21199.

**Pull request closed without merging**